### PR TITLE
Issue655

### PR DESCRIPTION
**Pull request by filippovdenis**
Compression feature is added. You can set compression_level option choosing one of the values: 'default_compression', 'best_compression', 'best_speed'. There is also 'no_compression' value when compression isn't needed which is used by default.

- [ ] tests added
- [ ] tests passing
- [ ] README updated (if needed)
- [ ] README Table of Contents updated (if needed)
- [ ] History.md and `version` in gemspec are untouched
- [ ] backward compatible
- [ ] feature works in `elasticsearch_dynamic` (not required but recommended)

### CODE REVIEW

**Comment by cosmo0920** [comment]
Thanks for scintillating your work!
But, current patch is conflicted due to big patch (#659).

Could you rebase of master?

File: lib/fluent/plugin/out_elasticsearch.rb [hunk 1]
```diff
@@ -309,6 +311,18 @@ class << self
           alias_method :split_request?, :split_request_size_check?
         end
       end
+
+      version_arr = Elasticsearch::Transport::VERSION.split('.')
+
+      if (version_arr[0].to_i < 7) || (version_arr[0].to_i == 7 && version_arr[1].to_i < 2)
+        if compression
+          raise Fluent::ConfigError, <<-EOC
+            Cannot use compression with elasticsearch-transport plugin version < 7.2.0
```
Comment (cosmo0920) [inline lib/fluent/plugin/out_elasticsearch.rb]: It is good for new users. Nice. :+1:

File: lib/fluent/plugin/out_elasticsearch.rb [hunk 2]
```diff
@@ -420,7 +451,10 @@ def client(host = nil)
         if local_reload_connections && @reload_after > DEFAULT_RELOAD_AFTER
           local_reload_connections = @reload_after
         end
-        headers = { 'Content-Type' => @content_type.to_s }.merge(@custom_headers)
+
+        gzip_headers = if compression then {'Content-Encoding' => 'gzip'} else {} end
```
Comment (cosmo0920) [inline lib/fluent/plugin/out_elasticsearch.rb]: Could you break single `if .. else .. end` line into multiple lines?

File: lib/fluent/plugin/out_elasticsearch_dynamic.rb [hunk 3]
```diff
@@ -44,21 +44,24 @@ def client(host = nil)
       @_es ||= begin
         @current_config = connection_options[:hosts].clone
         adapter_conf = lambda {|f| f.adapter @http_backend, @backend_options }
+        gzip_headers = if compression then {'Content-Encoding' => 'gzip'} else {} end
```
Comment (cosmo0920) [inline lib/fluent/plugin/out_elasticsearch_dynamic.rb]: ditto.

File: lib/fluent/plugin/out_elasticsearch.rb [hunk 4]
```diff
@@ -309,6 +311,18 @@ class << self
           alias_method :split_request?, :split_request_size_check?
         end
       end
+
+      version_arr = Elasticsearch::Transport::VERSION.split('.')
+
+      if (version_arr[0].to_i < 7) || (version_arr[0].to_i == 7 && version_arr[1].to_i < 2)
```
Comment (cosmo0920) [inline lib/fluent/plugin/out_elasticsearch.rb]: Can we use `Elasticsearch::Transport::VERSION < "7.2.0"` here?

Comment (filippovdenis) [reply lib/fluent/plugin/out_elasticsearch.rb]: Nope
When Transport version will be 10.0.0 it won't work, we'll get an error

Comment (cosmo0920) [reply lib/fluent/plugin/out_elasticsearch.rb]: Oh, I've missed the pitfall. Thanks. :smile:

**Pull request closed without merging**